src/cascadia/Remoting/Monarch.cpp

```diff
@@ -107,6 +108,35 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
         }
     }
 
+    // Method Description:
+    // - Gives the host process an opportunity to run any pre-close logic then
+    //   requests all peasants to close.
+    // Arguments:
+    // - <none> used
+    // Return Value:
+    // - <none>
+    void Monarch::_handleQuitAll(const winrt::Windows::Foundation::IInspectable& /*sender*/,
+                                 const winrt::Windows::Foundation::IInspectable& /*args*/)
+    {
+        // Let the process hosting the monarch run any needed logic before
+        // closing all windows.
+        _QuitAllRequestedHandlers(*this, nullptr);
+
+        // Tell all peasants to exit.
+        const auto callback = [&](const auto& /*id*/, const auto& p) {
+            p.Quit();
+        };
+        const auto onError = [&](const auto& id) {
+            TraceLoggingWrite(g_hRemotingProvider,
+                              "Monarch_handleQuitAll_Failed",
+                              TraceLoggingInt64(id, "peasantID", "The ID of the peasant which we could not close"),
+                              TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
+                              TraceLoggingKeyword(TIL_KEYWORD_TRACE));
+        };
+
+        _forEachPeasant(callback, onError);

```

Comment:
Just curious, so if all windows get told to `quit` at roughly the same time, including the monarch, would it pose any issues if the monarch was the first to quit while all other peasants are in the middle of quitting (maybe if some peasants had more tabs and took longer to quit)?

Reply:
I'm not sure. Since ultimately each of the `quit` actions is asynchronous (somewhere in the chain is a fire_and_forget method) it is hard to say. I imagine there might be some fighting for who becomes the monarch but I think that shouldn't matter since the other thread will continue quitting even while that happens. I haven't attempted to prove that behavior, so take what I just said with a grain of salt.

Reply:
As a test I opened ~100 windows and initiated `quit`. All of the windows closed, but there remained a "runtime broker" process which exited a few seconds afterwards. I don't know what the "runtime broker" is or if I should be concerned that it hung around on its own for a little bit.

Reply:
@leonMSFT With further testing, I was able to sometimes get it to close every window except one, that presumably became the new monarch, so I'll need to figure out how to fix that.